application/controllers/Hook.php

```diff
@@ -0,0 +1,254 @@
+        return $token == getenv('QANB_TOKEN');
+    }
+
+    /**
+     * Loop through data, recursive function
+     * @param $suite
+     * @param null $parent_suite_id
+     */
+    private function loopThrough($suite, $parent_suite_id = null) {
+
+        $data_suite = [
+            'execution_id' => $this->execution_id,
+            'uuid' => $suite->uuid,
+            'title' => $suite->title,
+            'campaign' => $this->extractNames($suite->file, 'campaign'),
+            'file' => $this->extractNames($suite->file, 'file'),
+            'duration' => $suite->duration,
+            'hasSkipped' => $suite->hasSkipped ? 1 :0,
+            'hasPasses' => $suite->hasPasses ? 1 :0,
+            'hasFailures' => $suite->hasFailures ? 1 :0,
+            'totalSkipped' => $suite->totalSkipped,
+            'totalPasses' => $suite->totalPasses,
+            'totalFailures' => $suite->totalFailures,
+            'hasSuites' => $suite->hasSuites ? 1 :0,
+            'hasTests' => $suite->hasTests ? 1 :0,
+            'parent_id' => $parent_suite_id,
+        ];
+
+
+        //inserting current suite
+        $suite_id = $this->Suite->insert($data_suite);
+
+        if ($suite_id) {
+            //insert tests
+            if (count($suite->tests) > 0) {
+                foreach($suite->tests as $test) {
+                    $data_test = [
+                        'suite_id' => $suite_id,
+                        'uuid' => $test->uuid,
+                        'title' => $test->title,
+                        'state' => $this->getTestState($test),
+                        'duration' => $test->duration,
+                        'error_message' => isset($test->err->message) ? $this->sanitize($test->err->message) : null,
+                        'stack_trace' => isset($test->err->estack) ? $this->sanitize($test->err->estack) : null,
+                        'diff' => isset($test->err->diff) ? $this->sanitize($test->err->diff) : null,
+                    ];
+
+                    $this->Test->insert($data_test);
+                }
+            }
+            //insert children suites
+            if (count($suite->suites) > 0) {
+                foreach($suite->suites as $s) {
+                    $this->loopThrough($s, $suite_id);
+                }
+            }
+        } else {
+            //we don't want to abort, just log this
+            log_message("error", "Error inserting suite into database.");
+        }
+    }
+
+    /**
+     * Sanitize text by removing weird characters
+     * @param $text
+     * @return string
+     */
+    private function sanitize($text) {
+        $StrArr = str_split($text);
+        $NewStr = '';
+        foreach ($StrArr as $Char) {
+            $CharNo = ord($Char);
+            if ($CharNo == 163) { $NewStr .= $Char; continue; }
+            if ($CharNo > 31 && $CharNo < 127) {
+                $NewStr .= $Char;
+            }
+        }
+        return $NewStr;
+    }
```

Comment:
Couldn't this be done with php's sanitization methods? ex:
```php
filter_var($text, FILTER_SANITIZE_STRING);
```

Reply:
TBH this method is a mess but I had a hard time making it work so I'm kinda afraid to touch it now